Re-run meta@name=color-scheme processing when name attribute or content attribute changes

sideshowbarker opened the pull request:
#### 3afefcc5f9348463406812ccd2ffab3dae4749a4
<pre>
Re-run meta@name=color-scheme processing when name attribute or content attribute changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=261654">https://bugs.webkit.org/show_bug.cgi?id=261654</a>

Reviewed by Aditya Keerthi.

This change makes WebKit conform to the requirement at
<a href="https://html.spec.whatwg.org/#standard-metadata-names">https://html.spec.whatwg.org/#standard-metadata-names</a>:attr-meta-name-15,
which states that:

&gt; [if] existing meta elements have their name or content attributes
changed, user agents must re-run the above algorithm

…where “above algorithm” is the algorithm at
<a href="https://html.spec.whatwg.org/#standard-metadata-names">https://html.spec.whatwg.org/#standard-metadata-names</a>:page&apos;s-supported-color-schemes-2

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-meta-element/color-scheme/meta-color-scheme-attribute-changes-expected.txt:
* Source/WebCore/html/HTMLMetaElement.cpp:
(WebCore::isNameColorScheme):
(WebCore::HTMLMetaElement::attributeChanged):
(WebCore::HTMLMetaElement::removedFromAncestor):
(WebCore::HTMLMetaElement::process):
* Source/WebCore/html/HTMLMetaElement.h:

Canonical link: <a href="https://commits.webkit.org/270092@main">https://commits.webkit.org/270092@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/643ff57070b04730f7e7bf2eedbef568e2774ca5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23074 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1157 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24190 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24989 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21340 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23340 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2438 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23610 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22197 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23315 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/633 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20010 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25840 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/499 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20896 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27066 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20894 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21160 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24933 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/579 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18363 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/507 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5878 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/988 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/780 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLMetaElement.cpp:
```diff
@@ -106,11 +106,19 @@ void HTMLMetaElement::attributeChanged(const QualifiedName& name, const AtomStri
+#if ENABLE(DARK_MODE_CSS)
+        if (const AtomString& nameValue = attributeWithoutSynchronization(nameAttr); equalLettersIgnoringASCIICase(nameValue, "color-scheme"_s) || equalLettersIgnoringASCIICase(nameValue, "supported-color-schemes"_s))
+            document().metaElementColorSchemeChanged();
+#endif
```
This code is sort of duplicated with what's inside the `process()` function. Except that method calls the outdated `document().processColorScheme()` method (which should probably be cleaned up/removed)

pxlcoder replied:
Following [268064@main](https://commits.webkit.org/268064@main), `process()` calls `metaElementColorSchemeChanged()`, so I'm wondering why this part of the change is necessary?

sideshowbarker replied:
> This code is sort of duplicated with what's inside the `process()` function. Except that method calls the outdated `document().processColorScheme()` method

`process()` on the main branch (and in this PR branch too) no longer calls `document().processColorScheme()` directly, right? Instead `process()` also now calls `metaElementColorSchemeChanged()`

>  (which should probably be cleaned up/removed)

When I earlier looked at `document().processColorScheme()` and `processColorSchemeString()`(and talked to you about them), it had seemed to me they were doing some parsing that was redundant with `CSSParser::parseSingleValue(CSSPropertyColorScheme,…)` — but after looking back at them further, I can see now that they’re not really re-parsing the value but instead are splitting the parsed color scheme (in the case where the color scheme contains multiple values), and then doing necessary processing based on the value.

But maybe I’m misunderstanding something?

sideshowbarker replied:
> Following [268064@main](https://commits.webkit.org/268064@main), `process()` calls `metaElementColorSchemeChanged()`, so I'm wondering why this part of the change is necessary?

(Apologies in advance if I’m misunderstanding which exact part you mean, but to be clear, I’ll attempt to write out here a prose explanation of what the code is attempting to do, and why)

`process()` doesn’t handle the case where the `name` or `content` attributes are _removed_; `process()` only looks at the _existing_ values of those attributes that it finds in the current DOM.

So, to describe explicitly the two things this patch is doing:
- In the case where a `name` attribute has changed — including the case where the `name` attribute has been completely removed (which `process()` doesn’t handle) — this patch checks the _old_ value of the `name` attribute, and if that _was_ `color-scheme`, then calls `document().metaElementColorSchemeChanged()`.
- In the case where a `content` attribute has changed — including the case where the `content` attribute has been completely removed (which `process()` doesn’t handle) — this patch checks the current value of the `name` attribute, and if that is `color-scheme`, then  calls `document().metaElementColorSchemeChanged()`.

If it seems like it might help to add comments to the code, please let me know, and I’ll do that.

In terms of test cases, without this patch, the following two asserts from http://wpt.live/html/semantics/document-metadata/the-meta-element/color-scheme/meta-color-scheme-attribute-changes.html fail:
- _Removed name attribute from meta color-scheme_
- _Removed the content attribute of meta color-scheme_

…Whereas, with this patch, those two asserts now pass.

sideshowbarker replied:
@pxlcoder, @nt1m Please lemme know if there’s anything more I need to do here at this point

pxlcoder replied:
> `process()` doesn’t handle the case where the name or content attributes are removed; `process()` only looks at the existing values of those attributes that it finds in the current DOM.

I think we might want to restructure that behavior? Currently, this code would result in scenarios where `metaElementColorSchemeChanged()` is called twice.

The comment in `process()` says: "// All below situations require a content attribute (which can be the empty string)." But that doesn't actually appear to be true for all of them.

sideshowbarker replied:
> Currently, this code would result in scenarios where `metaElementColorSchemeChanged()` is called twice.

Ah yeah — OK, I’ve pushed an update which attempts to address that. Please let me know if you don’t think it’s the right way to deal with it

pxlcoder commented on Source/WebCore/html/HTMLMetaElement.cpp:
```diff
@@ -106,11 +106,19 @@ void HTMLMetaElement::attributeChanged(const QualifiedName& name, const AtomStri
             if (equalLettersIgnoringASCIICase(oldValue, "theme-color"_s) && !equalLettersIgnoringASCIICase(newValue, "theme-color"_s))
                 document().metaElementThemeColorChanged(*this);
+#if ENABLE(DARK_MODE_CSS)
+            else if (equalLettersIgnoringASCIICase(oldValue, "color-scheme"_s) || equalLettersIgnoringASCIICase(oldValue, "supported-color-schemes"_s))
```
It might be time to factor this into a method, since it already appears twice in this file.

sideshowbarker replied:
> It might be time to factor this into a method, since it already appears twice in this file.

Thanks — the update I’ve pushed includes factoring it into a new function (which is a member function because — to deal with preventing the code from getting called twice — also sets a private data member).

pxlcoder commented on Source/WebCore/html/HTMLMetaElement.cpp:
```diff
@@ -192,6 +209,10 @@ void HTMLMetaElement::process()
         document().quirks().setNeedsToCopyUserSelectNoneQuirk();
+#if ENABLE(DARK_MODE_CSS)
+    else
+        maybeCallMetaElementColorSchemeChanged(document(), nameValue);
```
I would suggest updating the order of logic in `::process` to allow for a null `content`/`name` when processing `color-scheme`. This is similar to how the `http_equivAttr` is checked before the `name` null-check.

The current approach is more difficult to follow due to the use of `maybe`. It also increases the size of all `<meta>` elements with a new member, and I think that could be avoided.

pxlcoder replied:
That's what I had in mind with https://github.com/WebKit/WebKit/pull/17845#discussion_r1341696812, my apologies if I wasn't clear enough.

sideshowbarker replied:
OK, thanks much — I updated the order of logic in `::process`. But note that I also updated the parameter list for `::process` itself, to pass in the old value of the `name` attribute when it changes. Otherwise, without doing that, I can’t see how `::process` could handle the `name`-removal case.

I did also add a `::isNameColorScheme` member function — to avoid the repetition/verbosity of `equalLettersIgnoringASCIICase(nameValue, "color-scheme"_s) || equalLettersIgnoringASCIICase(nameValue, "supported-color-schemes"_s`.

cdumez commented on Source/WebCore/html/HTMLMetaElement.h:
```diff
@@ -50,11 +50,15 @@ class HTMLMetaElement final : public HTMLElement {
     void removedFromAncestor(RemovalType, ContainerNode&) final;
 
-    void process();
+    void process(const AtomString* oldValue = nullptr);
```
`const AtomString& oldValue = nullAtom()` then you can check `oldValue.isNull()` in the implementation. AtomString has a null state so we don't need to use a raw pointer here.

cdumez commented on Source/WebCore/html/HTMLMetaElement.h:
```diff
@@ -50,11 +50,15 @@ class HTMLMetaElement final : public HTMLElement {
     std::optional<Color> m_contentColor;
+
+#if ENABLE(DARK_MODE_CSS)
+    bool isNameColorScheme(const AtomString& nameValue);
```
Can be marked as `static`.

cdumez replied:
You don't even need to declare it in the header, having the static function at the top of the cpp file would be better here I think. If you insist on keeping it in the header, at least move it *before* the data members. Data members always come last, per our usual style.

sideshowbarker replied:
Thanks — marked `isNameColorScheme()` as `static` and moved it to the top of the cpp file

sideshowbarker committed:
Re-run meta@name=color-scheme processing when name attribute or content attribute changes
https://bugs.webkit.org/show_bug.cgi?id=261654

Reviewed by Aditya Keerthi.

This change makes WebKit conform to the requirement at
https://html.spec.whatwg.org/#standard-metadata-names:attr-meta-name-15,
which states that:

> [if] existing meta elements have their name or content attributes
changed, user agents must re-run the above algorithm

…where “above algorithm” is the algorithm at
https://html.spec.whatwg.org/#standard-metadata-names:page's-supported-color-schemes-2

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-meta-element/color-scheme/meta-color-scheme-attribute-changes-expected.txt:
* Source/WebCore/html/HTMLMetaElement.cpp:
(WebCore::isNameColorScheme):
(WebCore::HTMLMetaElement::attributeChanged):
(WebCore::HTMLMetaElement::removedFromAncestor):
(WebCore::HTMLMetaElement::process):
* Source/WebCore/html/HTMLMetaElement.h:

Canonical link: https://commits.webkit.org/270092@main
